File: src/skillguard/engines/structural/cross_platform_analyzer.py
"""Cross-platform reuse analysis engine.

Analyzes skill files for OWASP AST10 (Cross-Platform Reuse) security gaps,
detecting platform-specific patterns that may not work across different
AI coding assistant platforms.
"""

from __future__ import annotations

import re
import time

from skillguard.core.models import (
    DetectionRule,
    EngineResult,
    EngineVerdict,
    FileType,
    Finding,
    Severity,
    SkillFile,
)
from skillguard.engines.base import ScanEngine

# Cross-platform detection rules: (rule_id, rule_name, pattern, severity, confidence, description, owasp_ast, mitre_attack, remediation)
_XPLAT_RULES: list[dict] = [
    {
        "rule_id": "SG-XPLAT-001",
        "rule_name": "Platform-Specific API Without Guard",
        "pattern": r"(?i)(Bash\s*\(|execute_command|run_terminal|computer_use|browser_use)",
        "severity": Severity.MEDIUM,
        "confidence": 0.65,
        "description": (
            "Detected platform-specific tool call. These are Claude Code/Desktop "
            "specific APIs that won't work on other platforms."
        ),
        "owasp_ast": ["AST10"],
        "mitre_attack": [],
        "remediation": "Wrap platform-specific API calls in platform detection guards.",
    },
    {
        "rule_id": "SG-XPLAT-002",
        "rule_name": "Incompatible Permission Model",
        "pattern": r"(?i)(allow_commands|server_commands|mcp_servers|tool_permissions)",
        "severity": Severity.MEDIUM,
        "confidence": 0.60,
        "description": (
            "Detected platform-specific permission declaration in frontmatter/config. "
            "Permission models differ across AI coding platforms."
        ),
        "owasp_ast": ["AST10"],
        "mitre_attack": [],
        "remediation": "Use a universal permission manifest format compatible across platforms.",
    },
    {
        "rule_id": "SG-XPLAT-003",
        "rule_name": "Hardcoded Platform Path",
        "pattern": r"(?i)(~/\.claude/|~/\.cursor/|~/\.windsurf/|~/\.openclaw/|~/\.config/claude|\.claude/settings)",
        "severity": Severity.LOW,
        "confidence": 0.80,
        "description": (
            "Detected hardcoded platform-specific path. These paths are tied to a "
            "single platform and will break on others."
        ),
        "owasp_ast": ["AST10"],
        "mitre_attack": [],
        "remediation": "Use platform-agnostic path resolution instead of hardcoded platform paths.",
    },
    {
        "rule_id": "SG-XPLAT-004",
        "rule_name": "Transport Mismatch",
        "pattern": r"(?i)(transport:\s*stdio|transport:\s*sse|StdioServerTransport|SSEServerTransport)",
        "severity": Severity.MEDIUM,
        "confidence": 0.70,
        "description": (
            "Detected transport-specific code. When a skill assumes a specific transport, "
            "it may not work on platforms that use a different one."
        ),
        "owasp_ast": ["AST10"],
        "mitre_attack": [],
        "remediation": "Support both stdio and SSE transport modes, or document transport requirements.",
    },
    {
        "rule_id": "SG-XPLAT-005",
        "rule_name": "Sandbox Escalation Risk",
        "pattern": r"(?i)(host_mode:\s*true|sandbox:\s*false|disable_sandbox|no.?sandbox|unrestricted_mode)",
        "severity": Severity.HIGH,
        "confidence": 0.70,
        "description": (
            "Detected sandbox-bypassing feature. This feature is sandboxed on some "
            "platforms but unrestricted on others, creating a security gap."
        ),
        "owasp_ast": ["AST10"],
        "mitre_attack": ["T1190"],
        "remediation": "Ensure consistent sandboxing across all target platforms.",
    },
]


class CrossPlatformAnalyzer(ScanEngine):
    """Analyzes skills for cross-platform reuse security gaps (OWASP AST10)."""

    @property
    def name(self) -> str:
        return "cross_platform_analyzer"

    @property
    def version(self) -> str:
        return "0.4.0"

    async def scan(
        self,
        skill_files: list[SkillFile],
        rules: list[DetectionRule] | None = None,
    ) -> EngineResult:
        start = time.monotonic()
        findings: list[Finding] = []

        for sf in skill_files:
            if sf.content is None:
                continue

            file_findings = self._analyze_file(sf)
            findings.extend(file_findings)

        elapsed_ms = int((time.monotonic() - start) * 1000)

        if not findings:
            return EngineResult(
                engine_name=self.name,
                engine_version=self.version,
                verdict=EngineVerdict.CLEAN,
                confidence=0.7,
                findings=[],
                duration_ms=elapsed_ms,
            )

        max_confidence = max(f.confidence for f in findings)
        severities = {f.severity for f in findings}
        if Severity.HIGH in severities:
            verdict = EngineVerdict.SUSPICIOUS
        else:
            verdict = EngineVerdict.CLEAN

        return EngineResult(
            engine_name=self.name,
            engine_version=self.version,
            verdict=verdict,
            confidence=max_confidence,
            detection_name="Cross-Platform Analysis" if findings else None,
            findings=findings,
            duration_ms=elapsed_ms,
        )

    async def health_check(self) -> bool:
        return True

    def _analyze_file(self, sf: SkillFile) -> list[Finding]:
        """Scan a single file against all cross-platform rules."""
        findings: list[Finding] = []
        content = sf.content or ""

        for rule in _XPLAT_RULES:
            matches = list(re.finditer(rule["pattern"], content))
            for match in matches:
                line_start = content[: match.start()].count("\n") + 1
                snippet = match.group(0)

                findings.append(
                    Finding(
                        rule_id=rule["rule_id"],
                        rule_name=rule["rule_name"],
                        severity=rule["severity"],
                        category="cross_platform",
                        description=rule["description"],
                        file_path=sf.path,
                        line_start=line_start,
                        snippet=snippet,
                        owasp_ast=rule["owasp_ast"],
                        mitre_attack=rule["mitre_attack"],
                        confidence=rule["confidence"],
                        remediation=rule["remediation"],
                    )
                )

        return findings

File: go-skillguard/internal/engines/cross_platform_analyzer.go
package engines

import (
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/girdav01/skillguard/internal/core"
)

type xplatPattern struct {
	Name        string
	Pattern     *regexp.Regexp
	Severity    core.Severity
	Desc        string
	OWASPAST    []string
	MITREAttack []string
}

var xplatPatterns []xplatPattern

func init() {
	rawXplat := []struct {
		name, pattern string
		severity      core.Severity
		desc          string
		owaspAST      []string
		mitreAttack   []string
	}{
		{"Platform-Specific API", `(?i)(Bash\s*\(|execute_command|run_terminal|computer_use|browser_use)`, core.SeverityMedium, "Platform-specific API usage detected that may not be portable across agent platforms.", []string{"AST10"}, nil},
		{"Hardcoded Platform Path", `(?i)(~/\.claude/|~/\.cursor/|~/\.windsurf/|~/\.openclaw/)`, core.SeverityLow, "Hardcoded platform-specific path detected that reduces portability.", []string{"AST10"}, nil},
		{"Transport Mismatch", `(?i)(transport:\s*stdio|transport:\s*sse|StdioServerTransport|SSEServerTransport)`, core.SeverityMedium, "Transport protocol specification detected — verify compatibility across target platforms.", []string{"AST10"}, nil},
		{"Sandbox Escalation Risk", `(?i)(host_mode:\s*true|sandbox:\s*false|disable_sandbox|no.?sandbox|unrestricted_mode)`, core.SeverityHigh, "Sandbox escape or disable directive detected that could compromise platform security.", []string{"AST10"}, []string{"T1190"}},
	}
	for _, rx := range rawXplat {
		compiled, err := regexp.Compile(rx.pattern)
		if err != nil {
			continue
		}
		xplatPatterns = append(xplatPatterns, xplatPattern{rx.name, compiled, rx.severity, rx.desc, rx.owaspAST, rx.mitreAttack})
	}
}

// CrossPlatformAnalyzer detects platform-specific patterns and portability issues.
type CrossPlatformAnalyzer struct{}

func NewCrossPlatformAnalyzer() *CrossPlatformAnalyzer { return &CrossPlatformAnalyzer{} }

func (c *CrossPlatformAnalyzer) Name() string    { return "cross_platform_analyzer" }
func (c *CrossPlatformAnalyzer) Version() string { return "0.4.0" }

func (c *CrossPlatformAnalyzer) Scan(skillFiles []core.SkillFile, rules []core.DetectionRule) (*core.EngineResult, error) {
	start := time.Now()
	var findings []core.Finding

	for _, sf := range skillFiles {
		if sf.Content == nil {
			continue
		}
		content := *sf.Content
		for i, xp := range xplatPatterns {
			matches := xp.Pattern.FindAllStringIndex(content, -1)
			for _, match := range matches {
				lineStart := strings.Count(content[:match[0]], "\n") + 1
				snippet := content[match[0]:match[1]]
				if len(snippet) > 300 {
					snippet = snippet[:300]
				}
				ruleID := fmt.Sprintf("SG-XPLAT-%03d", i+1)
				findings = append(findings, core.Finding{
					RuleID:      ruleID,
					RuleName:    xp.Name,
					Severity:    xp.Severity,
					Category:    "cross_platform",
					Description: xp.Desc,
					FilePath:    sf.Path,
					LineStart:   &lineStart,
					Snippet:     &snippet,
					OWASPAST:    xp.OWASPAST,
					MITREAttack: xp.MITREAttack,
					Confidence:  0.75,
				})
			}
		}
	}

	elapsed := time.Since(start).Milliseconds()

	if len(findings) == 0 {
		return &core.EngineResult{
			EngineName:    c.Name(),
			EngineVersion: c.Version(),
			Verdict:       core.EngineVerdictClean,
			Confidence:    0.8,
			Findings:      []core.Finding{},
			DurationMs:    elapsed,
		}, nil
	}

	verdict := computeVerdict(findings)
	det := "Cross-Platform Issue"

	return &core.EngineResult{
		EngineName:    c.Name(),
		EngineVersion: c.Version(),
		Verdict:       verdict,
		Confidence:    computeConfidence(findings),
		DetectionName: &det,
		Findings:      findings,
		DurationMs:    elapsed,
	}, nil
}

func (c *CrossPlatformAnalyzer) HealthCheck() bool { return true }
